packages/core/src/amazonq/lsp/lspClient.ts

```diff
@@ -0,0 +1,219 @@
+ * Code reference: https://github.com/aws/language-servers/blob/7da212185a5da75a72ce49a1a7982983f438651a/client/vscode/src/credentialsActivation.ts#L77
+ */
+export function writeEncryptionInit(stream: Writable): void {
+    const request = {
+        version: '1.0',
+        mode: 'JWT',
+        key: key.toString('base64'),
+    }
+    stream.write(JSON.stringify(request))
+    stream.write('\n')
+}
+/**
+ * LspClient manages the API call between VS Code extension and LSP server
+ * It encryptes the payload of API call.
+ */
+export class LspClient {
+    static #instance: LspClient
+    client: LanguageClient | undefined
+
+    public static get instance() {
+        return (this.#instance ??= new this())
+    }
+
+    constructor() {
+        this.client = undefined
+    }
+
+    async encrypt(payload: string) {
+        return await new jose.CompactEncrypt(new TextEncoder().encode(payload))
+            .setProtectedHeader({ alg: 'dir', enc: 'A256GCM' })
+            .encrypt(key)
+    }
+
+    async indexFiles(request: string[], rootPath: string, refresh: boolean) {
+        try {
+            const encryptedRequest = await this.encrypt(
+                JSON.stringify({
+                    filePaths: request,
+                    rootPath: rootPath,
+                    refresh: refresh,
+                })
+            )
+            const resp = await this.client?.sendRequest(IndexRequestType, encryptedRequest)
+            return resp
+        } catch (e) {
+            getLogger().error(`LspClient: indexFiles error: ${e}`)
+            return undefined
+        }
+    }
+
+    async query(request: string) {
+        try {
+            const encryptedRequest = await this.encrypt(
+                JSON.stringify({
+                    query: request,
+                })
+            )
+            const resp = await this.client?.sendRequest(QueryRequestType, encryptedRequest)
+            return resp
+        } catch (e) {
+            getLogger().error(`LspClient: query error: ${e}`)
+            return []
+        }
+    }
+
+    async getLspServerUsage(): Promise<Usage | undefined> {
+        if (this.client) {
+            return (await this.client.sendRequest(GetUsageRequestType, '')) as Usage
+        }
+    }
+
+    async updateIndex(filePath: string) {
+        try {
+            const encryptedRequest = await this.encrypt(
+                JSON.stringify({
+                    filePath: filePath,
+                })
+            )
+            const resp = await this.client?.sendRequest(UpdateIndexRequestType, encryptedRequest)
+            return resp
+        } catch (e) {
+            getLogger().error(`LspClient: updateIndex error: ${e}`)
+            return undefined
+        }
+    }
+}
+/**
+ * Activates the language server, this will start LSP server running over IPC protocol.
+ * It will create a output channel named Amazon Q Language Server.
+ * This function assumes the LSP server has already been downloaded.
```

Comment:
`This function assumes the LSP server has already been downloaded.`, should we add a sanity check on this?

Reply:
this activate function is only called after the download is complete.